runtime/compiler/env/J9ClassEnv.cpp

```diff
@@ -417,24 +418,81 @@ J9::ClassEnv::isAnonymousClass(TR::Compilation *comp, TR_OpaqueClassBlock *clazz
    return comp->fej9()->isAnonymousClass(clazz);
    }
 
-const TR::TypeLayout*
-J9::ClassEnv::enumerateFields(TR::Region& region, TR_OpaqueClassBlock * opaqueClazz, TR::Compilation *comp)
-   {  
-   J9Class *clazz = (J9Class*)opaqueClazz;
-   TR_VMFieldsInfo fieldsInfo(comp, clazz, 1, stackAlloc);
-   ListIterator<TR_VMField> iter(fieldsInfo.getFields());
-   TR::TypeLayoutBuilder tlb(region);
-   for (TR_VMField *field = iter.getFirst(); field; field = iter.getNext())
+/*
+ * \param prefix
+ *    prefix is ended with `.`
+ */
+static char * mergeFieldNames(char * prefix, char * fieldName, bool withDotAtTheEnd, TR::Region &region)
+   {
+   int prefixLength = prefix ? strlen(prefix) : 0;
+   int nameLength = strlen(fieldName);
+   int mergedLength = nameLength + prefixLength;
+
+   if (withDotAtTheEnd)
+      mergedLength++;
+   mergedLength++; /* for adding \0 at the end */;
+
+   char * newName = new (region) char[mergedLength];
```

Comment:
I have this nagging concern about the number of allocations of `char` arrays and string copies that are happening to create the prefixes as things go along.  If you picture the recursion through the fields as a tree, there are allocations and copies of the prefix for each internal node of the tree, as well as for each leaf node.  I know the copies for the leaf nodes are required, but the copies for the internal nodes feel like overhead.

But to avoid all of that, I think the code would need to allocate a buffer for the maximum prefix length, or pass around an array of pointers to the pieces of the prefix, copying only at the leaf nodes.  And that would only avoid the copying for the internal nodes in the tree, not the leaf nodes.

So maybe the extra complication isn't worth it, but it might be something to keep in mind or at least run an experiment. . . .

Reply:
I think what you want here is a [Radix Tree](https://en.wikipedia.org/wiki/Radix_tree).

Moving to a clean Radix Tree representation is doable but will require a lot of work. To keep the complexity contained, we would have to refactor all the APIs that currently use strings (at least that ones related to type signatures and symrefs) to work in terms of tree nodes.

Reply:
I wasn't thinking of going as far as that.  I was thinking purely in the context of the one time creation of the flattened field names.  If there was a prepass, say, that determined the maximum length of the flattened and merged field names, a prefix buffer could be allocated.  The name of the "root" field could be copied into the prefix buffer just once, each of its child field's names could be copied into the prefix buffer just once, etc.

Whether any of this is really worth it really depends on to what extent we might have large value types that are nested "several" levels deep.  Maybe it's not worth considering right now, so I will withdraw this comment.

A simple small improvement that is worth making is adding a `prefixLength` argument to `mergeFieldNames`.  At least that would avoid the repeated calls to `strlen(prefix)`, when that is already known by the caller.

Reply:
> A simple small improvement that is worth making is adding a prefixLength argument to mergeFieldNames

I'll take a look at how to improve it. `addEntryForField()` is invoked recursively, so is `mergeFieldNames()`. To avoid the call of `strlen(prefix)` inside `mergeFieldNames()`, on top of my head I think the calculated length would need to be passed as a parameter through `addEntryForField()`  and `mergeFieldNames()` , and `mergeFieldNames()` needs to return the new length back for the child field.